pkg/peertls/extensions.go

```diff
@@ -20,59 +62,296 @@ type TLSExtConfig struct {
-					ok, err := v.f(ext, parsedChains)
+					err := v.f(ext, parsedChains)
 					if err != nil {
-						return ErrExtension.Wrap(utils.CombineErrors(v.err, err))
-					} else if !ok {
-						return v.err
+						return ErrExtension.Wrap(err)
 					}
 				}
 			}
 		}
 		return nil
 	}
 }
+
+func (r RevocationDB) Get(chain []*x509.Certificate) (*Revocation, error) {
+	hash, err := hashBytes(chain[CAIndex].Raw)
+	if err != nil {
+		return nil, ErrRevocation.Wrap(err)
+	}
+
+	revBytes, err := r.DB.Get(hash)
+	if err != nil && !storage.ErrKeyNotFound.Has(err) {
+		return nil, ErrRevocationDB.Wrap(err)
+	}
+	if revBytes == nil {
+		return nil, nil
+	}
+
+	rev := new(Revocation)
+	if err = rev.Unmarshal(revBytes); err != nil {
+		return rev, ErrRevocationDB.Wrap(err)
+	}
+	return rev, nil
+}
+
+func (r RevocationDB) Put(chain []*x509.Certificate, revExt pkix.Extension) error {
+	ca := chain[CAIndex]
+	var rev Revocation
+	if err := rev.Unmarshal(revExt.Value); err != nil {
+		return err
+	}
+
+	// TODO: what happens if cert/timestamp/sig is empty/garbage?
+
+	if err := rev.Verify(ca); err != nil {
+		return err
+	}
+
+	lastRev, err := r.Get(chain)
+	if err != nil {
+		return err
+	} else if lastRev != nil && lastRev.Timestamp >= rev.Timestamp {
+		return ErrRevocationTimestamp
+	}
+
+	hash, err := hashBytes(ca.Raw)
+	if err := r.DB.Put(hash, revExt.Value); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r Revocation) Verify(signingCert *x509.Certificate) error {
+	pubKey, ok := signingCert.PublicKey.(crypto.PublicKey)
+	if !ok {
+		return ErrUnsupportedKey.New("%T", signingCert.PublicKey)
+	}
+
+	data, err := r.TBSBytes()
+	if err != nil {
+		return err
+	}
+
+	if err := VerifySignature(r.Signature, data, pubKey); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *Revocation) TBSBytes() ([]byte, error) {
+
+	toHash := new(bytes.Buffer)
+	_, err := toHash.Write(r.CertHash)
+	if err != nil {
+		return nil, ErrExtension.Wrap(err)
+	}
+
+	// NB: append timestamp to revoked cert bytes
+	binary.PutVarint(toHash.Bytes(), r.Timestamp)
+
+	return hashBytes(toHash.Bytes())
+}
+
+func (r *Revocation) Sign(key crypto.PrivateKey) error {
+	data, err := r.TBSBytes()
+	if err != nil {
+		return err
+	}
+
+	r.Signature, err = signHashOf(key, data)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r Revocation) Marshal() ([]byte, error) {
+	data := new(bytes.Buffer)
+	// NB: using gob instead of asn1 because we plan to leave tls and asn1 is meh
+	encoder := gob.NewEncoder(data)
+	err := encoder.Encode(r)
+	if err != nil {
+		return nil, err
+	}
+
+	return data.Bytes(), nil
+}
+
+func (r *Revocation) Unmarshal(data []byte) error {
+	// NB: using gob instead of asn1 because we plan to leave tls and asn1 is meh
+	decoder := gob.NewDecoder(bytes.NewBuffer(data))
+	if err := decoder.Decode(r); err != nil {
+		return err
+	}
+	return nil
+}
+
+func verifyCAWhitelistSignedLeafFunc(caWhitelist []*x509.Certificate) extensionVerificationFunc {
```

Comment:
the reason it's at the bottom is because it's not an exported function but it's extension-specific so it didn't seem right to move it into utils. I've been sorting code like this:
1. constants
1. vars
1. exported functions
1. exported methods
1. unexported stuff

Honestly i'm not sure if that's "correct" or if it's in the best practices anywhere... I'm open to change tho.